commands/general/stats.js

```diff
@@ -4,118 +4,165 @@ const moment = require('moment')
-        mem = client.guilds.get(message.guild.id).members.get(userId)
-        if (message.author.avatarURL != null) {
-          av = message.author.avatarURL
+          userInfo.currentSession = 0
+          userInfo.overallSession = 0
+          userInfo.rank = `?? / ${tot}`
         }
+      } catch (err) {
+        console.log(err.stack)
+        return client.errorMessage(message,
+          `Error fetching stats for ${userInfo.username}#${userInfo.discriminator}: ${err}`)
       }
 
-      client.loadUserData(userId, async res => {
-        if (res === null) {
-          client.errorMessage(message, 'Error fetching your data from our servers, please try again.')
-          return
-        }
-
-        // set "semi-global" variables
-        let avatar
-        let source
-        let poss
-
-        // get leaderboard pos
-        await client.fetchWeeklyLeaderboardPos(message.guild.id, userId, pos => {
-          poss = pos.replace(/`/g, '')
+      render(userInfo).then(buffer => {
+        message.channel.send({
+          files: [{
+            attachment: buffer,
+            name: 'level.jpg'
+          }]
         })
-
-        // load template
-        await jimp.read('./assets/time_card.png')
-          .then(i => {
-            source = i
+        // delete response after set time
+          .then(m => {
+            setTimeout(() => {
+              m.delete()
+            }, client.settings.res_destruct_time * 1000)
           })
+      })
+    }
+  }
 
-        // overlay avatar on template image
-        await jimp.read(av)
-          .then(i => {
-            avatar = i
-          })
-        await avatar.resize(98, 98)
-        await source.composite(avatar, 14, 14)
+  /*
+   * The command works by building a userInfo structure with enough
+   * information to render a stats display for the user.
+   *
+   * userInfo = {
+   *   // via selectTargetUser
+   *   username,
+   *   discriminator,
+   *   // via selectTargetUser -> enrichUserData
+   *   userId,
+   *   mem,    // A discord.js GuildMember, with s_time
+   *   av,     // Avatar URL (or path)
+   *   // Our data
+   *   currentSession,
+   *   overallSession,
+   *   rank
+   * }
+   */
 
-        // check if the user is actively pracking and update times live if necessary
-        let activeTime = 0
-        let guild = await client.loadGuildData(message.guild.id)
-        // these last two conditions should be equivalent but maybe they were already pracking when the bot came up
-        if (mem.voiceChannel != null && guild.permitted_channels.includes(mem.voiceChannel.id) && !mem.mute && mem.s_time != null) {
-          activeTime = moment().unix() - mem.s_time
-        }
+  function selectTargetUser (message) {
+    let args = message.content.split(' ').splice(1)
+    let userInfo
+    if (args.length >= 1) {
+      userInfo = parseUserInfo(args)
+      if (userInfo === null) {
+        throw new Error('Unable to parse as username#discriminator.')
```

Comment:
Hmm. These are user errors, but the calling function logs any error thrown by us here. I do want to keep logging errors if we don't know why they're happening, but I'm not sure about putting user error conditions into the log - I feel like we should be able to go into the log and see any stack trace as a condition we haven't yet considered (i.e. a bug).